### PR TITLE
refactor update-anime feature

### PR DESCRIPTION
**Pull request by EsamSharaf**
PR pf a branch updates previous feature branch feature/update-anime-partially

### CODE REVIEW

File: anime/views.py [hunk 1]
```diff
@@ -45,31 +46,40 @@ def get_anime_by_name(name: str):
...
+
     try:
-        request_data = request.get_json()
-        db.session.query(Anime).filter(Anime.anime_id == id).update(
-            {**request_data}
-        )
-        db.session.commit()
+        schema_dict = anime_schema.loads(request.data)
+    except ValidationError as e:
+        return {"error": e.messages}, 400
```
Comment (blancheta) [inline anime/views.py:67]: error handler

Comment (blancheta) [reply anime/views.py:67]: """
# make sure the id exists in the database

# update the data
anime_to_update = Anime(**request_data)
db.save(anime_to_update)
db.session.commit()

return 
"""

https://stackoverflow.com/questions/60545807/python-sqlalchemy-orm-update-row-with-instance-of-class

**Review by blancheta — COMMENTED** [review]
A bit more work

File: anime/views.py (hunk 1 above)
Comment (EsamSharaf) [reply anime/views.py:67]: db.save doesn't exist. I leaned to another implementation
https://stackoverflow.com/questions/60545807/python-sqlalchemy-orm-update-row-with-instance-of-class

File: anime/views.py [hunk 2]
```diff
@@ -62,24 +62,29 @@ def update_anime(id: int):
...
-        schema_dict = anime_schema.loads(request.data)
+        entry_dict = anime_schema.loads(request.data)
     except ValidationError as e:
         return {"error": e.messages}, 400
 
-    _x = db.session.query(Anime).filter(Anime.anime_id == id).update(
-        {**schema_dict}
-    )
+    # Function to convert ORM obj to dictionary
+    row2dict = lambda r: {
```
Comment (blancheta) [inline anime/views.py:70]: I think there is a more elegant way.
https://stackoverflow.com/questions/1958219/how-to-convert-sqlalchemy-row-object-to-a-python-dict

Comment (EsamSharaf) [reply anime/views.py:66]: I have tried it but does not work, plus many methods but nothing is working. The main task is to convert  sqlalchemy.engine.row.Row data class to dict., only the method with green tick in the link I shared works-which is what I have implemented.

Comment (EsamSharaf) [reply anime/views.py]: I think the issue solved. Please have a look on commit [7e7efc7](https://github.com/EsamSharaf/anime_API/pull/8/commits/7e7efc7cf43ad8046667e78f985880718c91dbd4)

**Review by blancheta — COMMENTED** [review]
Thanks for your perseverance ;)

I will do a bit of research to help you with the solution i have in mind.
Usually the as_dict or to_dict method is available at the schema level.

File: anime/views.py [hunk 3]
```diff
@@ -45,31 +45,39 @@ def get_anime_by_name(name: str):
...
-        anime = db.session.execute(
-                    db.select(animes_tab).filter_by(anime_id=id)).first()
+    if request.method == 'PUT':
         anime_schema = AnimeSchema()
+    else:
+        anime_schema = AnimeSchema(partial=True)
+
+    entry_dict = anime_schema.loads(request.data)
+
+    db_anime = db.session.query(Anime).filter(Anime.anime_id == id)
```
Comment (blancheta) [inline anime/views.py:65]: if you use get here instead would you get a integrity exception you can catch with a flask handler?

File: anime/schemas.py [hunk 4]
```diff
@@ -2,12 +2,12 @@
...
 AnimeSchema = Schema.from_dict(
     {
-        "anime_id": fields.Int(),
-        "name": fields.Str(),
-        "genre": fields.Str(),
-        "type": fields.Str(),
-        "episodes": fields.Int(),
-        "rating": fields.Float(),
-        "members": fields.Int(),
+        "anime_id": fields.Int(required=True, error_messages={"required": "anime_id is required."}),
```
Comment (blancheta) [inline anime/schemas.py:5]: I wonder if those error messages are needed. Looks repetitive. Any other way to do not repeat this part?

File: anime/views.py (hunk 3 above)
Comment (EsamSharaf) [reply anime/views.py:65]: The issue is the get method can't being used, that using it in line 65 makes db_anime without update method, which  is used in line 76 to update the anime in the DB.

**Pull request merged**